File: backend-legacy-python/app/vpn/protocols/hysteria2.py
"""Hysteria2 — UDP-based protocol (runs as separate binary, no xray inbound).

Supports FinalMask UDP obfuscation modes:
- salamander: default obfs (built-in to hy2)
- xdns: mask UDP as DNS traffic (xray-level, post-transport)
- xicmp: mask UDP as ICMP traffic (xray-level, post-transport)
- off: no FinalMask obfuscation
"""

from urllib.parse import quote

from app.config import get_settings

from .base import ClientLink, Protocol, ServerConfig, UserCredentials, XrayInbound

# FinalMask modes that require xray-level outbound wrapping
_FINALMASK_XRAY_MODES = {"xdns", "xicmp"}


class Hysteria2(Protocol):
    def __init__(self) -> None:
        s = get_settings()
        self._password = s.hy2_password
        self._obfs_password = s.hy2_obfs_password
        self._port = s.hysteria2_port
        self._finalmask_mode = s.finalmask_mode

    @property
    def name(self) -> str:
        return "hysteria2"

    @property
    def display_name(self) -> str:
        return "Hysteria2"

    @property
    def enabled(self) -> bool:
        return bool(self._password)

    def xray_inbounds(
        self, users: list[UserCredentials], short_ids: list[str]
    ) -> list[XrayInbound]:
        return []  # Separate binary, not xray

    def xray_outbounds(self) -> list[dict]:
        """Return FinalMask obfuscation outbound when xdns/xicmp mode is active.

        FinalMask is applied at the xray level (post-transport) to mask
        UDP packets leaving the server, making Hysteria2 traffic look like
        DNS or ICMP to DPI systems.
        """
        if self._finalmask_mode not in _FINALMASK_XRAY_MODES:
            return []
        return [
            {
                "protocol": "freedom",
                "tag": "hy2-finalmask",
                "streamSettings": {
                    "finalmask": {
                        "type": self._finalmask_mode,
                    }
                },
            }
        ]

    def xray_routing_rules(self) -> list[dict]:
        """Route Hysteria2 traffic through FinalMask outbound when active."""
        if self._finalmask_mode not in _FINALMASK_XRAY_MODES:
            return []
        return [
            {
                "type": "field",
                "inboundTag": ["hy2-in"],
                "outboundTag": "hy2-finalmask",
            }
        ]

    def client_links(
        self, user: UserCredentials, servers: list[ServerConfig]
    ) -> list[ClientLink]:
        if not self._password:
            return []
        links: list[ClientLink] = []
        for srv in servers:
            remark = f"{srv.flag} {srv.name} Hysteria2"
            uri = (
                f"hy2://{self._password}@{srv.host}:{self._port}"
                f"?insecure=1&sni=rutube.ru&obfs=salamander"
                f"&obfs-password={self._obfs_password}"
                f"#{quote(remark)}"
            )
            links.append(ClientLink(uri=uri, protocol="hysteria2", transport="udp", server_key=srv.key, remark=remark))
        return links

    def singbox_outbound(
        self, tag: str, server: ServerConfig, user: UserCredentials, **opts
    ) -> dict | None:
        if not self._password:
            return None
        return {
            "type": "hysteria2",
            "tag": tag,
            "server": server.host,
            "server_port": self._port,
            "password": self._password,
            "tls": {"enabled": True, "server_name": "rutube.ru", "insecure": True},
            "obfs": {"type": "salamander", "password": self._obfs_password},
        }

File: backend/crates/chameleon-vpn/src/protocols/hysteria2.rs
//! Hysteria2 — UDP protocol with FinalMask obfuscation.

use serde_json::json;
use chameleon_config::Settings;
use super::types::*;

pub struct Hysteria2 {
    password: String,
    obfs_password: String,
    port: u16,
    sni: String,
    finalmask_mode: String,
}

impl Hysteria2 {
    pub fn new(s: &Settings) -> Self {
        Self {
            password: s.hy2_password.clone(),
            obfs_password: s.hy2_obfs_password.clone(),
            port: s.hysteria2_port,
            sni: s.hy2_sni.clone(),
            finalmask_mode: s.finalmask_mode.clone(),
        }
    }
}

impl Protocol for Hysteria2 {
    fn name(&self) -> &str { "hysteria2" }
    fn display_name(&self) -> &str { "Hysteria2" }
    fn enabled(&self) -> bool { !self.password.is_empty() }

    fn xray_inbounds(&self, _users: &[UserCredentials], _short_ids: &[String]) -> Vec<XrayInbound> {
        vec![] // Separate binary
    }

    fn xray_outbounds(&self) -> Vec<serde_json::Value> {
        if !matches!(self.finalmask_mode.as_str(), "xdns" | "xicmp") { return vec![]; }
        vec![json!({"protocol": "freedom", "tag": "hy2-finalmask", "streamSettings": {"finalmask": {"type": self.finalmask_mode}}})]
    }

    fn xray_routing_rules(&self) -> Vec<serde_json::Value> {
        if !matches!(self.finalmask_mode.as_str(), "xdns" | "xicmp") { return vec![]; }
        vec![json!({"type": "field", "inboundTag": ["hy2-in"], "outboundTag": "hy2-finalmask"})]
    }

    fn client_links(&self, _user: &UserCredentials, servers: &[ServerConfig]) -> Vec<ClientLink> {
        if self.password.is_empty() { return vec![]; }
        servers.iter().map(|srv| {
            let remark = format!("{} {} Hysteria2", srv.flag, srv.name);
            let uri = format!(
                "hy2://{}@{}:{}?insecure=1&sni={}&obfs=salamander&obfs-password={}#{}",
                self.password, srv.host, self.port, self.sni, self.obfs_password, urlencoding::encode(&remark)
            );
            ClientLink { uri, protocol: "hysteria2".into(), transport: "udp".into(), server_key: srv.key.clone(), remark, is_relay: false }
        }).collect()
    }

    fn singbox_outbound(&self, tag: &str, server: &ServerConfig, _user: &UserCredentials, _opts: &OutboundOpts) -> Option<serde_json::Value> {
        if self.password.is_empty() { return None; }
        Some(json!({
            "type": "hysteria2", "tag": tag, "server": server.host, "server_port": self.port,
            "password": self.password,
            "tls": {"enabled": true, "server_name": self.sni, "insecure": true},
            "obfs": {"type": "salamander", "password": self.obfs_password},
        }))
    }
}
